Runtime error with python 3.10.0 - undefined symbol: _PyGen_Send

Installing using `pip` completes successfully, but trying to use the module fails.

```
>>> import geopandas
>>> world = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/chaz/.local/lib/python3.10/site-packages/geopandas/io/file.py", line 166, in _read_file
    _check_fiona("'read_file' function")
  File "/home/chaz/.local/lib/python3.10/site-packages/geopandas/io/file.py", line 80, in _check_fiona
    raise ImportError(
ImportError: the 'read_file' function requires the 'fiona' package, but it is not installed or does not import correctly.
Importing fiona resulted in: /home/chaz/.local/lib/python3.10/site-packages/fiona/ogrext.cpython-310-x86_64-linux-gnu.so: undefined symbol: _PyGen_Send
```

CentOS 7 64bit
Linux 3.10.0-1160.42.2.el7.x86_64
proj 8.1.1
gdal 3.3.2
fiona 1.8.20
python 3.10.0

I'm having a related problem on macOS 12.1:  

```
    ERROR: Command errored out with exit status 1:
     command: /Users/federico/.pyenv/versions/3.10.0/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/0w/n2d7fhzj4c513yhfm2ln7l940000gn/T/pip-install-oobw74sa/fiona_d20d8bc75e6744bfac9e9c79c1af676e/setup.py'"'"'; __file__='"'"'/private/var/folders/0w/n2d7fhzj4c513yhfm2ln7l940000gn/T/pip-install-oobw74sa/fiona_d20d8bc75e6744bfac9e9c79c1af676e/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/0w/n2d7fhzj4c513yhfm2ln7l940000gn/T/pip-record-lnu9hgeh/install-record.txt --single-version-externally-managed --compile --install-headers /Users/federico/.pyenv/versions/3.10.0/include/python3.10/fiona
         cwd: /private/var/folders/0w/n2d7fhzj4c513yhfm2ln7l940000gn/T/pip-install-oobw74sa/fiona_d20d8bc75e6744bfac9e9c79c1af676e/
    Complete output (78 lines):
    /Users/federico/.pyenv/versions/3.10.0/lib/python3.10/distutils/dist.py:274: UserWarning: Unknown distribution option: 'metadata_version'
      warnings.warn(msg)
    /Users/federico/.pyenv/versions/3.10.0/lib/python3.10/distutils/dist.py:274: UserWarning: Unknown distribution option: 'requires_python'
      warnings.warn(msg)
    /Users/federico/.pyenv/versions/3.10.0/lib/python3.10/distutils/dist.py:274: UserWarning: Unknown distribution option: 'requires_external'
      warnings.warn(msg)
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-12.1-x86_64-3.10
    creating build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/_loading.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/env.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/compat.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/io.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/drvsupport.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/session.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/inspector.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/__init__.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/logutils.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/crs.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/rfc3339.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/vfs.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/transform.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/collection.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/_show_versions.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/errors.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    copying ./fiona/path.py -> build/lib.macosx-12.1-x86_64-3.10/fiona
    creating build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/options.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/collect.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/distrib.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/env.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/insp.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/cat.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/__init__.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/ls.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/rm.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/dump.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/calc.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/filter.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/load.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/main.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/info.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/helpers.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    copying ./fiona/fio/bounds.py -> build/lib.macosx-12.1-x86_64-3.10/fiona/fio
    running build_ext
    building 'fiona.schema' extension
    creating build/temp.macosx-12.1-x86_64-3.10
    creating build/temp.macosx-12.1-x86_64-3.10/fiona
    clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/usr/local/Cellar/gdal/3.3.3/include -I/Users/federico/.pyenv/versions/3.10.0/include/python3.10 -c fiona/schema.c -o build/temp.macosx-12.1-x86_64-3.10/fiona/schema.o
    clang -bundle -undefined dynamic_lookup -L/usr/local/opt/readline/lib -L/usr/local/opt/readline/lib -L/Users/federico/.pyenv/versions/3.10.0/lib -L/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib -L/usr/local/opt/readline/lib -L/usr/local/opt/readline/lib -L/Users/federico/.pyenv/versions/3.10.0/lib -L/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib build/temp.macosx-12.1-x86_64-3.10/fiona/schema.o -L/usr/local/Cellar/gdal/3.3.3/lib -lgdal -o build/lib.macosx-12.1-x86_64-3.10/fiona/schema.cpython-310-darwin.so
    building 'fiona._transform' extension
    clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/usr/local/Cellar/gdal/3.3.3/include -I/Users/federico/.pyenv/versions/3.10.0/include/python3.10 -c fiona/_transform.cpp -o build/temp.macosx-12.1-x86_64-3.10/fiona/_transform.o -std=c++11
    clang++ -bundle -undefined dynamic_lookup -L/usr/local/opt/readline/lib -L/usr/local/opt/readline/lib -L/Users/federico/.pyenv/versions/3.10.0/lib -L/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib -L/usr/local/opt/readline/lib -L/usr/local/opt/readline/lib -L/Users/federico/.pyenv/versions/3.10.0/lib -L/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib build/temp.macosx-12.1-x86_64-3.10/fiona/_transform.o -L/usr/local/Cellar/gdal/3.3.3/lib -lgdal -o build/lib.macosx-12.1-x86_64-3.10/fiona/_transform.cpython-310-darwin.so
    building 'fiona._geometry' extension
    clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/usr/local/Cellar/gdal/3.3.3/include -I/Users/federico/.pyenv/versions/3.10.0/include/python3.10 -c fiona/_geometry.c -o build/temp.macosx-12.1-x86_64-3.10/fiona/_geometry.o
    fiona/_geometry.c:13026:19: error: implicit declaration of function '_PyGen_Send' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                ret = _PyGen_Send((PyGenObject*)yf, value == Py_None ? NULL : value);
                      ^
    fiona/_geometry.c:13026:17: warning: incompatible integer to pointer conversion assigning to 'PyObject *' (aka 'struct _object *') from 'int' [-Wint-conversion]
                ret = _PyGen_Send((PyGenObject*)yf, value == Py_None ? NULL : value);
                    ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    fiona/_geometry.c:13031:19: error: implicit declaration of function '_PyGen_Send' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                ret = _PyGen_Send((PyGenObject*)yf, value == Py_None ? NULL : value);
                      ^
    fiona/_geometry.c:13031:17: warning: incompatible integer to pointer conversion assigning to 'PyObject *' (aka 'struct _object *') from 'int' [-Wint-conversion]
                ret = _PyGen_Send((PyGenObject*)yf, value == Py_None ? NULL : value);
                    ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    fiona/_geometry.c:13115:19: error: implicit declaration of function '_PyGen_Send' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                ret = _PyGen_Send((PyGenObject*)yf, NULL);
                      ^
    fiona/_geometry.c:13115:17: warning: incompatible integer to pointer conversion assigning to 'PyObject *' (aka 'struct _object *') from 'int' [-Wint-conversion]
                ret = _PyGen_Send((PyGenObject*)yf, NULL);
                    ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    3 warnings and 3 errors generated.
    error: command '/usr/bin/clang' failed with exit code 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /Users/federico/.pyenv/versions/3.10.0/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/0w/n2d7fhzj4c513yhfm2ln7l940000gn/T/pip-install-oobw74sa/fiona_d20d8bc75e6744bfac9e9c79c1af676e/setup.py'"'"'; __file__='"'"'/private/var/folders/0w/n2d7fhzj4c513yhfm2ln7l940000gn/T/pip-install-oobw74sa/fiona_d20d8bc75e6744bfac9e9c79c1af676e/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/0w/n2d7fhzj4c513yhfm2ln7l940000gn/T/pip-record-lnu9hgeh/install-record.txt --single-version-externally-managed --compile --install-headers /Users/federico/.pyenv/versions/3.10.0/include/python3.10/fiona Check the logs for full command output.
```

I could reproduce the error on the `python:3.10.0-bullseye` docker image with Cython 0.29.21. Upgrading to a cython version > 0.29.21 seems to work fine.  I needed to upgrade the dev requirements as some seemed to be incompatible with Python 3.10:

Dockerfile:
```
FROM python:3.10.0-bullseye

# Install gdal
RUN apt-get update && \
    apt-get upgrade -y && \
    apt-get install -y gdal-bin libgdal-dev && \
    pip install --upgrade pip

COPY Fiona/requirements.txt .
COPY Fiona/requirements-dev.txt .

RUN pip install -r requirements-dev.txt

COPY Fiona /Fiona
WORKDIR /Fiona

RUN python -m pip install --no-deps --force-reinstall --no-use-pep517 -e .

RUN python -m pytest -m "not wheel" --cov fiona --cov-report term-missing
```

Updated requirements-dev.txt:

```
-r requirements.txt
coverage==6.1.2
cython==0.29.22
pytest==6.2.5
pytest-cov==3.0.0
setuptools==59.2.0
boto3==1.20.10
wheel==0.37.0
pytz==2021.3
```

With cython==0.29.21, `docker build` in the directory of the Dockerfile leads to the following error:

```
ImportError while loading conftest '/Fiona/tests/conftest.py'.
tests/conftest.py:13: in <module>
    import fiona
fiona/__init__.py:86: in <module>
    from fiona.collection import BytesCollection, Collection
fiona/collection.py:11: in <module>
    from fiona.ogrext import Iterator, ItemsIterator, KeysIterator
E   ImportError: /Fiona/fiona/ogrext.cpython-310-x86_64-linux-gnu.so: undefined symbol: _PyGen_Send
```

For cython>0.29.21, all tests run successfully.

--- a/.github/workflows/ci_linux.yml
+++ b/.github/workflows/ci_linux.yml
@@ -80,12 +80,18 @@ jobs:
               PROJVERSION: "7.2.1",
               allow_failure: "false",
             }
+          - {
+              python: "3.10",
+              GDALVERSION: "3.3.3",
+              PROJVERSION: "8.2.0",
+              allow_failure: "false",
+            }
 
           # Test GDAL master
           - {
-              python: 3.6,
+              python: 3.9,
               GDALVERSION: "master",
-              PROJVERSION: "7.2.1",
+              PROJVERSION: "8.2.0",
               allow_failure: "true",
             }
 

--- a/fiona/drvsupport.py
+++ b/fiona/drvsupport.py
@@ -101,7 +101,7 @@ supported_drivers = dict([
     # multi-layer
     #   ("OpenAir", "r"),
     # PCI Geomatics Database File 	PCIDSK 	No 	No 	Yes, using internal PCIDSK SDK (from GDAL 1.7.0)
-    ("PCIDSK", "rw"),
+    ("PCIDSK", "raw"),
     # PDS 	PDS 	No 	Yes 	Yes
     ("PDS", "r"),
     # PDS renamed to OGR_PDS for GDAL 2.x
@@ -162,6 +162,7 @@ driver_mode_mingdal = {
           'FlatGeobuf': (3, 1, 3)},
 
     'a': {'GPKG': (1, 11, 0),
+          'PCIDSK': (3, 3, 0),
           'GeoJSON': (2, 1, 0),
           'MapInfo File': (2, 0, 0)}
 }

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,2 +1,7 @@
 [build-system]
-requires = ["setuptools", "wheel", "cython==0.29.21", "oldest-supported-numpy"]
+requires = [
+    "cython==0.29.24",
+    "oldest-supported-numpy",
+    "setuptools",
+    "wheel",
+]

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -1,10 +1,9 @@
 -r requirements.txt
-pytest==5.0.1
-setuptools==41.6.0
-boto3==1.9.19
-coverage==4.5.4
-cython==0.29.21
-pytest-cov==2.8.1
-pytz==2020.1
-setuptools==41.6.0
-wheel==0.33.6
+pytest==6.2.5
+boto3==1.20.10
+coverage==6.1.2
+cython==0.29.24
+pytest-cov==3.0.0
+pytz==2021.3
+setuptools==58.5.3
+wheel==0.37.0

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,6 +1,6 @@
-attrs==18.2.0
+attrs==21.2.0
 click-plugins==1.0.4
 cligj==0.5.0
 munch==2.3.2
-six==1.11.0
+six==1.16.0
 certifi

--- a/setup.py
+++ b/setup.py
@@ -340,6 +340,8 @@ setup_args = dict(
         'Programming Language :: Python :: 3.6',
         'Programming Language :: Python :: 3.7',
         'Programming Language :: Python :: 3.8',
+        'Programming Language :: Python :: 3.9',
+        'Programming Language :: Python :: 3.10',
         'Topic :: Scientific/Engineering :: GIS'])
 
 if os.environ.get('PACKAGE_DATA'):

